lib/experimental/sync/class-gutenberg-http-signaling-server.php

```diff
@@ -0,0 +1,343 @@
+<?php
+/**
+ * Gutenberg_HTTP_Signaling_Server class
+ *
+ * @package    Gutenberg
+ */
+
+/**
+ * Gutenberg class that contains an HTTP based signaling server used for collaborative editing.
+ *
+ * @access private
+ * @internal
+ */
+class Gutenberg_HTTP_Signaling_Server {
+
+	/**
+	 * Contains the path of the subscriber to messages file.
+	 * The file contains a data-structure similar to the following example:
+	 * array( 2323232121 => array( 'message hello','handshake message' ) ).
+	 *
+	 * @access private
+	 * @var array
+	 */
+	private static $subscriber_to_messages_path;
+
+	/**
+	 * Contains the path of the topics to subscriber file.
+	 * The file contains a data-structure similar to the following example:
+	 * array( 'doc1: array( 2323232121 ), 'doc2: array( 2323232123, 2323232121 ) ).
+	 *
+	 * @access private
+	 * @var array
+	 */
+	private static $topics_to_subscribers_path;
+
+	/**
+	 * Contains the path of the subscribers to last connection file.
+	 * The file contains a data-structure similar to the following example:
+	 * array( 2323232121 => 34343433323(timestamp) ).
+	 *
+	 * @access private
+	 * @var array
+	 */
+	private static $subscribers_to_last_connection_path;
+
+	/**
+	 * Contains the subscriber id of the client reading or sending messages.
+	 *
+	 * @access private
+	 * @var Integer
+	 */
+	private static $subscriber_id;
+
+	/**
+	 * Adds a wp_ajax action to handle the signaling server requests.
+	 */
+	public static function init() {
+		add_action( 'wp_ajax_gutenberg_signaling_server', array( __CLASS__, 'do_wp_ajax_action' ) );
+	}
+
+	/**
+	 * Handles a wp_ajax signaling server request.
+	 */
+	public static function do_wp_ajax_action() {
```

Comment:
What would you think of organizing the server code by the available operations? Being `do_wp_ajax_action` the main method, we could have the following methods:

- `handle_subscribe_to_topics`
- `handle_unsubscribe_from_topics`
- `handle_publish_message`
- `handle_read_pending_messages`
- `handle_ping`

The `handle_*` methods would take as input all the data they need, no access to global data (subscriber_id, REQUEST), so they are pure and easy to test. All the REQUEST related logic would be left to `do_wp_ajax_action`, as it is now.

Reply:
The code was refactored to follow this logic.